radio/src/storage/sdcard_yaml.cpp

```diff
@@ -47,23 +49,69 @@ const char * readYamlFile(const char* fullpath, const YamlParserCalls* calls, vo
     if (result != FR_OK) {
         return SDCARD_ERROR(result);
     }
-      
+
     YamlParser yp; //TODO: move to re-usable buffer
     yp.init(calls, parser_ctx);
 
+    uint16_t calculated_checksum = 0xFFFF;
+    uint16_t file_checksum = 0;
+    bool first_block = true;
     char buffer[32];
     while (f_read(&file, buffer, sizeof(buffer), &bytes_read) == FR_OK) {
 
       // reached EOF?
       if (bytes_read == 0)
         break;
-      
+
+      uint16_t skip = 0;
+      if(first_block) {
+        // Get the 'checksum' value and skip from further YAML processing
+        // The checksum must be first in the first buffer read from file
+        first_block = false;
+        if(strncmp(buffer, "checksum: ", 10) == 0) {
+          skip = 10;
+          char* startPos = buffer + 10; // length of "checksum: "
+          char* endPos = startPos;
+          while(*endPos != '\r') {
+            if (endPos > buffer + bytes_read) {
+              return SDCARD_ERROR(	FR_INT_ERR);
+            }
+            endPos++;
+          }
+          *endPos = 0;
+
+          // Skip LF of line-ending CR-LF pair
+          endPos++;
+          if(*endPos == '\n') {
+            endPos++;
+          }
+          file_checksum = atoi(startPos);
+          skip = endPos - buffer;
+        }
+      }
+
+      // Calculate checksum on read block only if we are called with a pointer to write the resulting checksum
+      if (checksum_result != NULL) {
+        calculated_checksum = crc16(0, (const uint8_t *)buffer + skip, bytes_read - skip, calculated_checksum);
+      }
+
       if (f_eof(&file)) yp.set_eof();
-      if (yp.parse(buffer, bytes_read) != YamlParser::CONTINUE_PARSING)
+      if (yp.parse(buffer + skip, bytes_read - skip) != YamlParser::CONTINUE_PARSING)
         break;
     }
-
     f_close(&file);
+
+
+    if (checksum_result != NULL) {
+      if (calculated_checksum == file_checksum) {
+        *checksum_result = CHECKSUM_SUCCESS;
+      } else {
+        *checksum_result = CHECKSUM_FAILED;
+      }
+    } else {
+      *checksum_result = CHECKSUM_NONE;
+    }
```

Comment:
I think line 112 is not a good idea. Dereferencing a NULL pointer.

Reply:
Good catch

Reply:
Update on its way with null-pointer dereference fixed and enum class for return values.